Draft Nuclia Documentation (#1529)
plone/documentation: nuclia.md

> # Replacing Sphinx Search with Nuclia Search in Plone

Reviewer: ```suggestion
# Configuring Nuclia to search across multiple websites
```

> ## Overview

Reviewer: All headings should have content underneath them. This heading is unnecessary, and leaves the previous heading without content.

```suggestion
```

> This documentation outlines the process of replacing Sphinx search in a Plone site with Nuclia search. The goal is to explain the approach taken and provide instructions for replicating the process in another project.

Reviewer: This rewrite elaborates what you have done, why it is awesome, and serves as an introduction. I also added some [MyST syntax](https://6.docs.plone.org/contributing/documentation/myst-reference.html) that uses arbitrary cross-references to a [page](https://6.docs.plone.org/contributing/documentation/myst-reference.html#link-to-a-chapter-or-page) and using [Intersphinx](https://myst-parser.readthedocs.io/en/latest/syntax/cross-referencing.html#cross-project-intersphinx-links).

We prefer one sentence per line. See https://6.docs.plone.org/contributing/documentation/authors.html#plone-documentation-styleguide

```suggestion
This documentation outlines the process of replacing the default Sphinx search in documentation websites with Nuclia search.

With Sphinx search, you are limited to searching only that one site.
Nuclia's "knowledge box" can index content from multiple websites and display results across them.

The implementation described below displays search results across the {doc}`Plone 6 Documentation </index>` and `Training <inv:training:std:doc:index>` websites.
You can follow and replicate this process for your projects.
```

## Approach

> ### Data Indexing and Processing

Reviewer: Unnecessary heading, also leaves previous heading empty.

```suggestion
```

> To achieve the transition, the following steps were taken:

Reviewer: Change from passive voice to active voice, and address the reader in a friendly way.

```suggestion
To achieve the transition from Sphinx to Nuclia search, we took an approach as shown in the following generalized outline.
```

1. **Generate Nuclia Sync Data:**
   - Iterate through `.md` files in the repository [docs](https://github.com/plone/documentation/tree/6.0/docs) .
   - Calculate the hash of each file and store the mapping in `nuclia_sync.json`.

2. **Extract Headings:**
   - Use regular expressions to extract headings from Markdown files.
   - Store the extracted headings and slugs for document URLs.

3. **Upload Documents:**
   - Utilize the `NucliaUpload` class from the `nuclia` SDK.
   - Generated slugs, Extracted headings, Nuclia `API_KEY` and constructed URLs for `.md` can be passed in the class method.

4. **Synchronize Documents:**
   - Compare hashes to determine whether to upload, update, or delete documents, when changes are made.
   - Handle document deletions using the `NucliaKB` class.

5. **GitHub Actions Workflow:**
   - Created a synchronization workflow triggered by a `push` event.

>    - Defined the steps for checking out code, setting up the Python environment, running the sync script, and committing changes back to the repository.

Reviewer: To avoid dozens of changes, I'm consolidating all the changes for the above outline into one big suggestion. Note that I always use 4 spaces to indent things. See https://stackoverflow.com/a/48313531/2214933 and https://stackoverflow.com/a/45683081/2214933. I also simplified formatting, and made a few minor grammar changes.

```suggestion
1.  Generate Nuclia sync data

    -   Iterate through documentation source files in the repository [docs](https://github.com/plone/documentation/tree/6.0/docs).
    -   Calculate the hash of each file, and store the mapping in `nuclia_sync.json` at the repository root.

1.  Extract headings

    -   Use regular expressions to extract headings from documentation source files.
    -   Store the extracted headings and slugs for document URLs.

1.  Upload documents

    -   Use the `NucliaUpload` class from the `nuclia` SDK.
    -   Generated slugs, extracted headings, Nuclia `API_KEY`, and constructed URLs for documentation source files can be passed into the class method.

1.  Synchronize documents

    -   Compare hashes to determine whether to upload, update, or delete documents, when changes are made.
    -   Handle document deletions using the `NucliaKB` class.

1.  GitHub Actions workflow

    -   Create a synchronization workflow triggered by a `push` event.
    -   Define the steps for checking out code, setting up the Python environment, running the sync script, and committing changes back to the repository.
```

## Replication Steps

Follow these steps to replicate the process in your own project:

> ### Set Up

Reviewer: 4 spaces to indent things.

```suggestion

### Pre-requisites

-   See {ref}`installation` for requirements to build Plone 6 Documentation.
-   See `Training <inv:training:std:lable:installation>` for requirements to build Plone Training.
-   [insert requirements for Nuclia and anything else as additional list items]

```

Copy this in your terminal to clone plone/documentation

```bash
 git clone `<repository-url>`

 pip install -q -r requirements-initial.txt

 pip install -q -r requirements.txt

> ```

Reviewer: The above content should be removed, as it is superseded by the standard installation method.

> ### Global Configuration

Reviewer: ```suggestion
### Global configuration

```

1. **PUBLIC_URL:**
    - Set the URL of your website
2. **API Key:**
    - Obtain the API key from your nuclia Knowledge Box.
3. **Knowledge Base URL:**

>     - Set up the variable to define the Nuclia knowledge box URL.

Reviewer: Reformatting.

Where does the global config go?

Also is it "Knowledge Base URL" or "Knowledge Box URL"?

Is "knowledge box" a proper noun or copyrighted term? In other words, should it be Title Cased?

```suggestion
1.  `PUBLIC_URL`
    -   Set the URL of your website.
2.  API Key
    -   Obtain the API key from your Nuclia knowledge box.
3.  Knowledge Base URL
    -   Set up the variable to define the Nuclia knowledge box URL.
```

> ### Indexing and Syncing

Reviewer: ```suggestion
### Indexing and syncing

```

1. **Generate Nuclia Sync Data:**

    ```bash
    python3 upload.py
    ```

2. **Sync Documents:**

- Run the `sync` function to upload, update, or delete documents in Nuclia.

> 

Reviewer: Replace the above with the reformatted list:

```suggestion
1.  Generate Nuclia sync data.

    ```shell
    python3 upload.py
    ```

2.  Sync documents.
    -   Run the `sync` function to upload, update, or delete documents in Nuclia.
```

### GitHub Actions

- **Workflow Setup:**
     Modify the GitHub Actions workflow `nuclia_sync.yml` to match your repository structure.

>## Usage Notes
>
>- Ensure that the API key and knowledge base URL are correctly configured.
>- Regularly update the sync process to keep the knowledge base up to date.
>- Troubleshoot issues by checking API key validity and document URLs.

### Conclusion

Replacing Sphinx search with Nuclia search brings improved search functionality to Plone sites. Feel free to reach out for assistance or clarification on any aspect of this documentation.

> Happy syncing and searching!

Reviewer: Reformatting of the above section

```suggestion
Modify the GitHub Actions workflow `nuclia_sync.yml` to match your repository structure.
 
## Usage notes

-   Ensure that the API key and knowledge base URL are correctly configured.
-   Regularly update the sync process to keep the knowledge base up to date.
-   Troubleshoot issues by checking the API key validity and document URLs.

## Conclusion

Replacing Sphinx search with Nuclia search brings improved search functionality to Plone sites.
Feel free to reach out for assistance or clarification on any aspect of this documentation.
Happy syncing and searching!
